File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Views/List.kt
package com.abracode.actionui.Views

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.Button as M3Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text as M3Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.abracode.actionui.Common.ActionUIElement
import com.abracode.actionui.Common.ActionUILogger
import com.abracode.actionui.Common.ActionUIModel
import com.abracode.actionui.Common.ActionUIRegistry
import com.abracode.actionui.Common.ActionUIValueType
import com.abracode.actionui.Common.ActionUIViewConstruction
import com.abracode.actionui.Common.ContainerShape
import com.abracode.actionui.Common.LocalActionUILogger
import com.abracode.actionui.Common.LocalWindowModel
import com.abracode.actionui.Common.LoggerLevel
import com.abracode.actionui.Helpers.BuildViewWithModifiers
import com.abracode.actionui.Helpers.LocalActionUIInputEnabled
import com.abracode.actionui.Helpers.ProvideTextStyleEnvironment
import com.abracode.actionui.Helpers.RefreshableScrollContainer
import com.abracode.actionui.Helpers.RegisterLazyScrollHandle
import com.abracode.actionui.Helpers.TemplateHelper
import com.abracode.actionui.Helpers.applyScrollContentBackground
import com.abracode.actionui.Helpers.stringProperty
import com.abracode.actionui.Helpers.templateRows
import kotlinx.serialization.json.JsonObject

/**
 * Scrollable list. Mirror of the Apple `List` element (`ActionUI/Views/List.swift`),
 * which wraps `SwiftUI.List`. Like [LazyVStack] it maps to a Compose [LazyColumn]
 * - both the lazy builder and the scroll viewport in one - so it needs a bounded
 * height (an inherited `frame.height` wins, else [DEFAULT_MAIN_EXTENT]; the same
 * "always a finite viewport" stance the lazy stacks take).
 *
 * Three modes, resolved in the same precedence as Swift:
 *
 *   1. **Template (data-driven repeater)** - when a [ActionUIElement.template] is
 *      present: one substituted template instance per row in
 *      `states[`[ActionUIModel.ROWS_STATE_KEY]`]` (set via the rows API), with
 *      `$1`/`$2`/`$0` column substitution. See `Helpers/TemplateHelper.kt`.
 *   2. **Heterogeneous (children)** - when `children` are present: each child is a
 *      full ActionUI view, built through the registry. This is the only mode that
 *      renders from a static document (no host data needed).
 *   3. **Homogeneous (itemType)** - otherwise: each row renders its first column
 *      via [resolveItemType]. `Text` (default) and `Button` are supported;
 *      `Button` fires its `itemType.actionID` with the row's value (or index,
 *      when `actionContext` is `"rowIndex"`) as context.
 *
 * The rows are seeded empty via [initialStates] so a host can address them by id
 * before data arrives (the element needs a positive `id`, as on Apple).
 *
 * **Selection (value bridge, B6).** Like Apple, the selected row is exposed as the
 * element's [ActionUIValueType.STRING_LIST] `value` - the selected row's columns
 * (empty when nothing is selected) - so a host reads it with
 * `ActionUIModel.getElementValueAsString` (tab-joined) and writes it with
 * `setElementValueFromString`. Selection is **interactive only when a list-level
 * `actionID` is present** (Apple's selectable mode): in the data-row modes
 * (template / homogeneous) a tap selects the row, stores it as the value,
 * highlights it, and fires `actionID` with the row as `context`. Single-selection
 * only. Per-row `Button` actions still fire independently (the inner control
 * consumes its own tap).
 *
 * **Pull-to-refresh.** When the element carries an `onRefreshActionID`, the LazyColumn is
 * wrapped in a Material3 pull-to-refresh box (see [RefreshableScrollContainer]); a pull fires
 * the actionID and the indicator stays until the client delivers fresh rows. Apple parity is
 * the `.refreshable` modifier on the same element.
 *
 * **Deferred vs. Apple (documented).**
 *   * **Children-mode selection.** Heterogeneous `children` are arbitrary,
 *     possibly-interactive views, so wrapping each in a selection surface would
 *     fight their own taps (the conflict Apple sidesteps by requiring Label/Text
 *     children under `List(selection:)`). Children mode keeps its children
 *     interactive and is not row-selectable; the value bridge is still live, so a
 *     host can set/read the selection programmatically.
 *   * **`listStyle`** and the `listRow*` styling (background / separator / insets)
 *     are accepted in the schema but not honored here (no portable Compose
 *     `List` styling surface); they are ignored silently.
 *
 * Named `ListView` (not `List`) so the type does not shadow `kotlin.collections.List`
 * across the `Views` package; it is registered under the canonical string `"List"`.
 */
object ListView : ActionUIViewConstruction {

    override val insertableContainers = mapOf("children" to ContainerShape.FLAT)

    /** The selected row, as the Apple-parity `[String]` selection value. */
    override val valueType = ActionUIValueType.STRING_LIST

    /** Nothing selected initially (matches Apple's empty `[String]` default). */
    override fun initialValue(element: ActionUIElement): Any? = emptyList<String>()

    override fun initialStates(element: ActionUIElement): Map<String, Any> =
        mapOf(ActionUIModel.ROWS_STATE_KEY to emptyList<List<String>>())

    @Composable
    override fun BuildView(element: ActionUIElement, modifier: Modifier) {
        val logger = LocalActionUILogger.current
        val props = element.properties
        val template = element.template
        val children = element.children.orEmpty()

        // A LazyColumn is its own scroll viewport and needs a bounded height; an
        // inherited frame.height wins, else a default keeps an unbounded parent
        // from crashing it (same guard as LazyVStack).
        val hasExplicitHeight = (props?.get("frame") as? JsonObject)?.get("height") != null
        val listModifier = (if (hasExplicitHeight) modifier else modifier.height(DEFAULT_MAIN_EXTENT))
            .applyScrollContentBackground(props, logger)

        // Selection (value bridge): interactive only when a list-level actionID is
        // present (Apple's selectable mode). The selected row is the element value.
        val viewModel = LocalWindowModel.current?.viewModels?.get(element.id)
        val actionID = props?.stringProperty("actionID")
        val selectable = actionID != null
        @Suppress("UNCHECKED_CAST")
        val selection: List<String> = (viewModel?.value as? List<String>) ?: emptyList()
        val onSelect: (Int, List<String>) -> Unit = { index, row ->
            if (selection != row) {
                viewModel?.value = row
                actionID?.let {
                    ActionUIModel.actionHandler(it, viewID = element.id, viewPartID = index, context = row)
                }
            }
        }

        // Hoisted so an enclosing ScrollViewReader can drive scroll-to-row.
        // Only children mode is addressable (template / homogeneous rows
        // carry no element ids); see ScrollReaderHelper.kt.
        val listState = rememberLazyListState()
        RegisterLazyScrollHandle(element, listState, vertical = true)

        // A hidden subtree must not consume touch - Apple's `.hidden()` and web's `display:none`
        // are non-interactive, but Android `hidden` is only `alpha(0f)` (invisible, still laid out
        // AND still hit-testable). In the section-switcher idiom (overlapping bodies in a ZStack,
        // all but one hidden) an invisible List on top in z-order would otherwise steal the vertical
        // drag: its LazyColumn is hit first and Compose stops hit-testing there, so the visible List
        // behind never scrolls (Missing_Features #34). A disabled `userScrollEnabled` is not enough -
        // the LazyColumn is still a hit target and still blocks the sibling. So when hidden we render
        // a bare Box that reserves the same bounded viewport size (the reserve-space semantics stay,
        // Missing_Features #30) but carries no scrollable/pointer node, so hit-testing falls through
        // to the visible sibling behind. LocalActionUIInputEnabled is provided false for a hidden
        // subtree by ProvideReactiveEnvironment (see ReactiveEnvironment.kt).
        if (!LocalActionUIInputEnabled.current) {
            Box(listModifier)
            return
        }

        // Pull-to-refresh wraps the scroll viewport when onRefreshActionID is set; otherwise
        // this renders the LazyColumn unchanged (see RefreshableScrollContainer).
        RefreshableScrollContainer(element) {
            when {
                template != null -> {
                    val rows = templateRows(element.id)
                    LazyColumn(state = listState, modifier = listModifier) {
                        itemsIndexed(rows) { index, row ->
                            SelectableRow(selectable, selected = selection == row, onClick = { onSelect(index, row) }) {
                                TemplateHelper.BuildTemplateRow(
                                    template = template, row = row, parentID = element.id, rowIndex = index,
                                )
                            }
                        }
                    }
                }

                children.isNotEmpty() -> {
                    LazyColumn(state = listState, modifier = listModifier) {
                        items(children) { child ->
                            val builder = ActionUIRegistry.lookup(child.type) ?: return@items
                            ProvideTextStyleEnvironment(child.properties, logger) {
                                builder.BuildViewWithModifiers(child, Modifier)
                            }
                        }
                    }
                }

                else -> {
                    val rows = templateRows(element.id)
                    val itemType = resolveItemType(props, logger)
                    LazyColumn(state = listState, modifier = listModifier) {
                        itemsIndexed(rows) { index, row ->
                            SelectableRow(selectable, selected = selection == row, onClick = { onSelect(index, row) }) {
                                HomogeneousRow(itemType, row.firstOrNull().orEmpty(), index, element.id)
                            }
                        }
                    }
                }
            }
        }
    }

    /** Default scroll-viewport height when JSON supplies no `frame.height`. */
    private val DEFAULT_MAIN_EXTENT = 320.dp
}

/**
 * Wraps a data row so it can be selected. When [selectable] is false the [content]
 * renders bare (no selection affordance); otherwise it is wrapped in a
 * full-width, tappable surface that highlights when [selected]. An inner control
 * (e.g. a `Button` cell) consumes its own tap, so per-row actions and row
 * selection coexist.
 */
@Composable
private fun SelectableRow(
    selectable: Boolean,
    selected: Boolean,
    onClick: () -> Unit,
    content: @Composable () -> Unit,
) {
    if (!selectable) {
        content()
        return
    }
    val background = if (selected) MaterialTheme.colorScheme.secondaryContainer else Color.Transparent
    Box(Modifier.fillMaxWidth().background(background).clickable(onClick = onClick)) {
        content()
    }
}

/**
 * Resolved homogeneous `itemType`. [viewType] is narrowed to what the Android
 * renderer draws today - `"Text"` or `"Button"`; `Image`/`AsyncImage` cells
 * downgrade to `Text` (the image-resolution contract is the B2 track).
 */
internal data class ListItemType(
    val viewType: String,
    val actionContext: String,
    val actionID: String?,
)

/**
 * Resolves the `itemType` object for homogeneous mode, defaulting to a `Text`
 * cell and warning (matching Swift's validator) on an unknown `viewType` or
 * `actionContext`. `Image`/`AsyncImage` warn and render as text (B2). Pure
 * (logging aside) so it is unit-testable.
 */
internal fun resolveItemType(props: JsonObject?, logger: ActionUILogger): ListItemType {
    val itemType = props?.get("itemType") as? JsonObject

    val viewType = when (val raw = itemType?.stringProperty("viewType") ?: "Text") {
        "Text", "Button" -> raw
        "Image", "AsyncImage" -> {
            logger.log(
                "List itemType.viewType '$raw' renders as text on Android (image contract deferred, B2)",
                LoggerLevel.warning,
            )
            "Text"
        }
        else -> {
            logger.log("List itemType.viewType '$raw' invalid; defaulting to Text", LoggerLevel.warning)
            "Text"
        }
    }

    val actionContext = when (val raw = itemType?.stringProperty("actionContext")) {
        null, "title", "rowIndex" -> raw ?: "title"
        else -> {
            logger.log("List itemType.actionContext '$raw' invalid; defaulting to title", LoggerLevel.warning)
            "title"
        }
    }

    return ListItemType(viewType, actionContext, itemType?.stringProperty("actionID"))
}

/**
 * One homogeneous row. A `Text` cell shows the value; a `Button` cell fires its
 * `itemType.actionID` with the owning list id as `viewID`, the row index as
 * `viewPartID`, and the value (or the index, for `actionContext == "rowIndex"`)
 * as context.
 */
@Composable
private fun HomogeneousRow(itemType: ListItemType, value: String, index: Int, parentID: Int) {
    when (itemType.viewType) {
        "Button" -> M3Button(
            onClick = {
                itemType.actionID?.let { actionID ->
                    val context: Any = if (itemType.actionContext == "rowIndex") index else value
                    ActionUIModel.actionHandler(
                        actionID, viewID = parentID, viewPartID = index, context = context,
                    )
                }
            }
        ) { M3Text(text = value) }
        else -> M3Text(text = value)
    }
}

File: ActionUIWeb/src/Views/List.js
// List.js — List element.
// Web analog of ActionUI/Views/List.swift (and ActionUIAndroid Views/List.kt).
//
// A vertical collection of rows. Apple's List has three modes:
//
//   1. Homogeneous (itemType)  — one cell type per row, rows supplied as data.
//   2. Heterogeneous (children) — arbitrary child views, one per row.
//   3. Template (data-driven)  — one substituted template instance per row.
//
// Web ships all three modes. Mode precedence matches Swift / Android: template
// (when present) → children → homogeneous.
//
// Data-driven rows. Modes 1 and 3 read their rows from `states["content"]`
// ([[String]]) via the rows API on Window (set/append/clearElementRows) — the
// same store Table uses. The List binds its "content" state, so a host
// set/append/clear re-renders the rows in place. Homogeneous mode displays each
// row's **first column** (matching Apple/Android) as the `itemType.viewType`
// cell; the web renders real Image / AsyncImage cells (Android downgrades those
// to text, its B2 deferral). Template mode renders one substituted template
// instance per row ($1/$2/$N column references; see Helpers/TemplateHelper.js).
//
// Selection (value bridge). Apple exposes the selection as a `[String]`; the web
// value vocabulary is scalar, so — like Android's tab-joined string transport —
// the selection is carried as a **String**. In children mode it is the
// stringified id of the selected child; in the data-driven modes it is the
// selected row's columns tab-joined ("" = nothing selected). A host reads/writes
// it with get/setString(listID). A row can also be selected programmatically
// (without replacing the rows, firing no actionID) via the model selection API —
// selectElementRow(index) / selectElementRowWithContent(text, column) /
// clearElementSelection (the data-driven modes only; heterogeneous children select
// by child id). Selection is interactive only when a list-level
// `actionID` is present (Apple's selectable mode): clicking a row selects it,
// highlights it, and fires `actionID` (the host then reads the selection — Apple
// fires with no context and reads model.value, so the web matches and sends none).
// A row whose click lands on an interactive control (a Button cell) lets that
// control consume the tap, so per-row actions and row selection coexist — the web
// can disambiguate by event target, where Apple asks for Label/Text children.
//
// Properties (mirroring List.swift):
//   itemType         { viewType, actionContext, actionID, dataInterpretation } —
//                    homogeneous cell config.
//   actionID         Fires on selection change (enables selectable mode).
//   onRefreshActionID String; when set, enables pull-to-refresh: a downward pull at
//                    the top fires this actionID and an indicator stays up until the
//                    client delivers fresh rows to this list or anything inside it
//                    (the web side of SwiftUI's .refreshable; see PullToRefreshHelper).
//   doubleClickActionID  macOS double-click; row index as context (data modes).
//   listStyle        "automatic"|"plain"|"inset"|"sidebar" (the macOS set).
//   listRowBackground / listRowSeparator / listRowSeparatorTint / listRowInsets
//                    Row styling applied uniformly to every row.

import { register } from "../Common/ActionUIRegistry.js";
import { markHandlesAction, resolveColor } from "../Common/ModifierResolver.js";
import { buildDataImageCell } from "../Helpers/DataImageCell.js";
import { buildTemplateRow } from "../Helpers/TemplateHelper.js";
import { commonRowPrefix } from "../Helpers/RowDiff.js";
import { navigateRowsOnKey } from "../Helpers/RowKeyboardNav.js";
import { ContainerShape } from "../Common/ActionUIInsertion.js";
import { flatContainerBinding } from "../Helpers/InsertionHelper.js";

// The macOS-valid listStyles (the web's default skin is macOS-flavored). The
// other SwiftUI styles ("grouped", "insetGrouped") are iOS/tvOS/visionOS-only and
// warn here exactly as they do on macOS.
const VALID_LIST_STYLES = ["automatic", "plain", "inset", "sidebar"];
const ITEM_VIEW_TYPES = ["Text", "Button", "Image", "AsyncImage"];
const DATA_INTERPRETATIONS = ["path", "systemName", "assetName", "resourceName", "mixed"];
const ITEM_ACTION_CONTEXTS = ["title", "rowIndex"];

// Interactive descendants that should consume their own tap instead of selecting
// the row (so a Button cell fires its action while a Text cell selects the row).
const INTERACTIVE_SELECTOR = "button, input, select, textarea, a";

register("List", {
    // The selected child id / row, as a String ("" = nothing selected). See header.
    valueType: "string",

    // Opts into the pull-to-refresh build wrapper (Helpers/PullToRefreshHelper.js):
    // a no-op unless `onRefreshActionID` is set. Apple parity is `.refreshable`.
    refreshable: true,

    // Children mode accepts runtime insertions (insertElement / removeElement);
    // each inserted child is wrapped in a row and joins selection like the
    // statically authored rows. The data-driven modes mutate via the rows API
    // (set/append/clearElementRows), not this container.
    insertableContainers: { children: ContainerShape.FLAT },

    // Mirrors List.swift validateProperties (warning text included verbatim).
    validateProperties: (properties, logger) => {
        const validated = { ...properties };

        // itemType — normalized to an object with a valid viewType (Apple always
        // writes back a validated itemType, defaulting to Text).
        const itemType = (typeof properties.itemType === "object" && properties.itemType !== null)
            ? { ...properties.itemType } : { viewType: "Text" };
        const viewType = typeof itemType.viewType === "string" ? itemType.viewType : "Text";
        if (!ITEM_VIEW_TYPES.includes(viewType)) {
            logger.log("List itemType.viewType must be 'Text', 'Button', 'Image', or 'AsyncImage'; defaulting to Text", "warning");
            itemType.viewType = "Text";
        }
        if (viewType === "Image" && !DATA_INTERPRETATIONS.includes(itemType.dataInterpretation)) {
            logger.log(`List itemType.dataInterpretation must be 'path', 'systemName', 'assetName', 'resourceName', or 'mixed' for ${viewType}; defaulting to systemName`, "warning");
            itemType.dataInterpretation = "systemName";
        }
        if (viewType === "Button" && !ITEM_ACTION_CONTEXTS.includes(itemType.actionContext)) {
            logger.log("List itemType.actionContext must be 'title' or 'rowIndex' for Button; defaulting to title", "warning");
            itemType.actionContext = "title";
        }
        validated.itemType = itemType;

        // doubleClickActionID — must be a string, else dropped.
        if (validated.doubleClickActionID !== undefined && typeof validated.doubleClickActionID !== "string") {
            logger.log("List doubleClickActionID must be a string; ignoring", "warning");
            delete validated.doubleClickActionID;
        }

        // onRefreshActionID (pull-to-refresh) — must be a string, else dropped.
        if (validated.onRefreshActionID !== undefined && typeof validated.onRefreshActionID !== "string") {
            logger.log("List onRefreshActionID must be a string; ignoring", "warning");
            delete validated.onRefreshActionID;
        }

        // listStyle — must be a String valid on this (macOS-flavored) platform.
        if (typeof validated.listStyle === "string") {
            if (!VALID_LIST_STYLES.includes(validated.listStyle)) {
                logger.log(`List listStyle '${validated.listStyle}' is not available on this platform; ignoring`, "warning");
                delete validated.listStyle;
            }
        } else if (validated.listStyle !== undefined) {
            logger.log("List listStyle must be a String; ignoring", "warning");
            delete validated.listStyle;
        }

        // listRowBackground — must be a color string.
        if (validated.listRowBackground !== undefined && typeof validated.listRowBackground !== "string") {
            logger.log("List listRowBackground must be a color string; ignoring", "warning");
            delete validated.listRowBackground;
        }

        // listRowSeparator — must be one of visible/hidden/automatic.
        if (typeof validated.listRowSeparator === "string") {
            if (!["visible", "hidden", "automatic"].includes(validated.listRowSeparator)) {
                logger.log("List listRowSeparator must be 'visible', 'hidden', or 'automatic'; ignoring", "warning");
                delete validated.listRowSeparator;
            }
        } else if (validated.listRowSeparator !== undefined) {
            logger.log("List listRowSeparator must be a String; ignoring", "warning");
            delete validated.listRowSeparator;
        }

        // listRowSeparatorTint — must be a color string.
        if (validated.listRowSeparatorTint !== undefined && typeof validated.listRowSeparatorTint !== "string") {
            logger.log("List listRowSeparatorTint must be a color string; ignoring", "warning");
            delete validated.listRowSeparatorTint;
        }

        // listRowInsets — must be a number or an edge dictionary.
        if (validated.listRowInsets !== undefined) {
            const insets = validated.listRowInsets;
            const isNumber = typeof insets === "number";
            const isDict = typeof insets === "object" && insets !== null && !Array.isArray(insets);
            if (!isNumber && !isDict) {
                logger.log("List listRowInsets must be a number or dictionary {top, leading, bottom, trailing}; ignoring", "warning");
                delete validated.listRowInsets;
            }
        }

        return validated;
    },

    // Nothing selected initially (Apple's empty [String] → the empty string here).
    initialValue: () => "",

    // The documented row store (states["content"] = [[String]]); empty until a
    // host fills it via the rows API. Read by the data-driven modes.
    initialStates: () => ({ content: [] }),

    buildView: (element, properties, ctx) => {
        const node = document.createElement("div");
        node.className = "aui-list";
        if (typeof properties.listStyle === "string") {
            node.dataset.auiListStyle = properties.listStyle;
            node.classList.add(`aui-list-style-${properties.listStyle}`);
        }
        // The list owns its actionID semantics (selection), so it opts out of the
        // generic click→dispatch handler in applyViewModifiers.
        markHandlesAction(node);

        const children = element.children();
        const selectable = typeof properties.actionID === "string";
        const rowStyle = computeRowStyle(properties, ctx.logger);

        // Mode precedence (matching List.swift / List.kt): template → children →
        // homogeneous.
        const template = element.template();
        if (template) {
            // Template (data-driven repeater): one substituted instance per content
            // row, reusing the shared data-row plumbing.
            return buildDataRows(node, element, properties, ctx, selectable, rowStyle,
                (row, index) => buildTemplateRow(template, row, index, element.id, ctx));
        }

        if (children.length > 0) {
            return buildChildrenRows(node, element, properties, ctx, children, selectable, rowStyle);
        }

        // Homogeneous (itemType) mode: each row renders its first column as a
        // single typed cell. validateProperties has already normalized itemType.
        const itemType = (typeof properties.itemType === "object" && properties.itemType !== null)
            ? properties.itemType : { viewType: "Text" };
        return buildDataRows(node, element, properties, ctx, selectable, rowStyle,
            (row, index) => renderHomogeneousCell(itemType, row[0] ?? "", index, element, ctx));
    },
});

// Heterogeneous children mode: each child is a full ActionUI view built through
// the registry into a row. Selection (when the list has an actionID) is carried
// as the stringified selected child id. Selection styling reads the live rows
// (querySelectorAll), so a runtime-inserted child (insertElement) participates
// without extra bookkeeping.
function buildChildrenRows(node, element, properties, ctx, children, selectable, rowStyle) {
    if (selectable) node.setAttribute("role", "listbox");

    let selectedId = "";

    const applySelectionStyles = () => {
        for (const row of node.querySelectorAll(".aui-list-row")) {
            const isSelected = selectable && row.dataset.auiRowId === selectedId && selectedId !== "";
            row.classList.toggle("aui-list-row-selected", isSelected);
            if (selectable) row.setAttribute("aria-selected", isSelected ? "true" : "false");
        }
    };

    const selectRow = (id, fromUser) => {
        if (!selectable || id === selectedId) return;
        selectedId = id;
        applySelectionStyles();
        // Apple fires the selection-change actionID with no context and lets the
        // host read model.value; the web carries the selection as the element
        // value (getString), so it matches — no context sent.
        if (fromUser) ctx.model.dispatchAction(properties.actionID, element.id, 0, null);
    };

    // Wraps one built child view in a row + selection wiring. Shared by the
    // initial build and runtime insertion (the container binding's `wrap`).
    const wrapRow = (childNode, childId) => {
        const row = document.createElement("div");
        row.className = "aui-list-row";
        applyRowStyle(row, rowStyle);
        row.appendChild(childNode);
        if (selectable) {
            row.dataset.auiRowId = String(childId);
            row.setAttribute("role", "option");
            row.setAttribute("aria-selected", "false");
            row.tabIndex = 0;
            row.addEventListener("click", (event) => {
                // A click on an interactive cell (e.g. a Button) is the cell's
                // own; only a click on inert content selects the row.
                const control = event.target.closest?.(INTERACTIVE_SELECTOR);
                if (control && row.contains(control)) return;
                selectRow(row.dataset.auiRowId, true);
            });
            row.addEventListener("keydown", (event) => {
                if (event.key === "Enter" || event.key === " ") {
                    event.preventDefault();
                    selectRow(row.dataset.auiRowId, true);
                    return;
                }
                // Arrow / Home / End move selection+focus to the adjacent row. The
                // current position is resolved live (insertion can reorder `slots`).
                const i = slots.findIndex((slot) => slot.dom === row);
                navigateRowsOnKey(event, i, slots.length, (target) => {
                    selectRow(slots[target].dom.dataset.auiRowId, true); // string id, like the click path
                    slots[target].dom.focus();
                });
            });
        }
        return row;
    };

    const slots = children.map((child) => {
        const row = wrapRow(ctx.build(child), child.id);
        node.appendChild(row);
        return { id: child.id, dom: row };
    });

    if (element.id > 0) {
        ctx.model.bind(element.id, {
            getValue: () => selectedId,
            setValue: (value) => { // programmatic: silent (no actionID dispatch)
                selectedId = String(value ?? "");
                applySelectionStyles();
            },
        });
        // The insertion container: a new child is wrapped via wrapRow and tracked
        // by its element id (so before/after-sibling positions resolve).
        ctx.model.bindContainer(element.id, "children",
            flatContainerBinding(node, slots, { wrap: (childNode, id) => wrapRow(childNode, id) }));
    }

    return node;
}

// Shared plumbing for the data-driven row modes (homogeneous itemType now, and
// the template repeater next): rows come from states["content"], and the
// selection rides as the tab-joined row String. `renderCell(row, index)` builds
// the mode's row body; everything else (selection, styling, re-render on a rows
// change) is identical across the modes.
function buildDataRows(node, element, properties, ctx, selectable, rowStyle, renderCell) {
    if (selectable) node.setAttribute("role", "listbox");

    let rows = [];
    let rowNodes = [];
    let selectedRow = ""; // the selected row's columns tab-joined ("" = none)

    const rowValue = (row) => row.join("\t");

    const applySelectionStyles = () => {
        rowNodes.forEach((rowNode, i) => {
            const isSelected = selectable && selectedRow !== "" && rowValue(rows[i]) === selectedRow;
            rowNode.classList.toggle("aui-list-row-selected", isSelected);
            if (selectable) rowNode.setAttribute("aria-selected", isSelected ? "true" : "false");
        });
    };

    const selectRow = (index, fromUser) => {
        const value = rowValue(rows[index]);
        if (value === selectedRow) return;
        selectedRow = value;
        applySelectionStyles();
        // Apple fires the selection actionID with no context and reads
        // model.value; the web matches — the selection is the element value.
        if (fromUser) ctx.model.dispatchAction(properties.actionID, element.id, 0, null);
    };

    const buildRowNode = (row, index) => {
        const rowNode = document.createElement("div");
        rowNode.className = "aui-list-row";
        applyRowStyle(rowNode, rowStyle);
        rowNode.appendChild(renderCell(row, index));
        if (selectable) {
            rowNode.setAttribute("role", "option");
            rowNode.setAttribute("aria-selected", "false");
            rowNode.tabIndex = 0;
            rowNode.addEventListener("click", (event) => {
                const control = event.target.closest?.(INTERACTIVE_SELECTOR);
                if (control && rowNode.contains(control)) return;
                selectRow(index, true);
            });
            rowNode.addEventListener("keydown", (event) => {
                if (event.key === "Enter" || event.key === " ") {
                    event.preventDefault();
                    selectRow(index, true);
                    return;
                }
                // Arrow / Home / End move the selection (and focus) to the adjacent
                // row, the listbox keyboard pattern - else the key scrolls the parent.
                navigateRowsOnKey(event, index, rowNodes.length, (target) => {
                    selectRow(target, true);
                    rowNodes[target].focus();
                });
            });
            rowNode.addEventListener("dblclick", () => {
                if (typeof properties.doubleClickActionID === "string" && rowValue(row) === selectedRow) {
                    ctx.model.dispatchAction(properties.doubleClickActionID, element.id, 0, index);
                }
            });
        }
        return rowNode;
    };

    // Apply a new rows array with a common-prefix diff (Helpers/RowDiff.js) instead
    // of a full rebuild: an append (the rows API re-sends the whole array) keeps
    // every unchanged prefix row - its node, selection styling and wiring - in place
    // and builds only the new tail; a prepend / mid-list edit rebuilds from the first
    // change (which re-bakes the shifted per-row indices in the click/dblclick
    // handlers). The diff preserves the selection across an append.
    const applyRows = (next) => {
        next = Array.isArray(next) ? next : [];
        const keep = commonRowPrefix(rows, next);
        while (rowNodes.length > keep) { rowNodes[rowNodes.length - 1].remove(); rowNodes.pop(); }
        for (let index = keep; index < next.length; index++) {
            const rowNode = buildRowNode(next[index], index);
            rowNodes.push(rowNode);
            node.appendChild(rowNode);
        }
        rows = next;
        node.classList.toggle("aui-list-empty", rows.length === 0);
        // A rows change that drops the selected row clears the selection.
        if (selectedRow !== "" && !rows.some((row) => rowValue(row) === selectedRow)) selectedRow = "";
        applySelectionStyles();
    };

    if (element.id > 0) {
        // The rows store: set/append/clearElementRows route through here and
        // re-render the rows in place (incrementally - see applyRows).
        ctx.model.bindState(element.id, {
            getState: (key) => (key === "content" ? rows : undefined),
            setState: (key, value) => { if (key === "content") applyRows(value); },
        });
        // The selection value: the selected row tab-joined; highlighted silently
        // when set programmatically.
        ctx.model.bind(element.id, {
            getValue: () => selectedRow,
            setValue: (value) => {
                selectedRow = String(value ?? "");
                applySelectionStyles();
            },
        });
    }

    applyRows([]);
    return node;
}

// One homogeneous itemType cell: the row's first column rendered as the
// itemType.viewType. Text shows the value; Button fires itemType.actionID
// (context = the row index or the value) and consumes its tap; Image / AsyncImage
// render a real image (via the shared data-image-cell helper).
function renderHomogeneousCell(itemType, value, index, element, ctx) {
    const viewType = itemType.viewType ?? "Text";
    if (viewType === "Button") {
        const button = document.createElement("button");
        button.className = "aui-list-button";
        button.textContent = value;
        button.addEventListener("click", (event) => {
            event.stopPropagation(); // the cell's own tap, not row selection
            if (typeof itemType.actionID === "string") {
                const context = itemType.actionContext === "rowIndex" ? index : value;
                ctx.model.dispatchAction(itemType.actionID, element.id, 0, context);
            }
        });
        return button;
    }
    if (viewType === "Image") {
        return buildDataImageCell(value, itemType.dataInterpretation ?? "systemName", ctx.logger, "List");
    }
    if (viewType === "AsyncImage") {
        const img = document.createElement("img");
        img.className = "aui-data-image";
        img.loading = "lazy";
        img.src = value;
        img.alt = "";
        return img;
    }
    const span = document.createElement("span");
    span.textContent = value;
    return span;
}

// Resolves the row-styling properties once; the result is applied to every row.
function computeRowStyle(properties, logger) {
    const style = {};
    if (typeof properties.listRowBackground === "string") {
        style.background = resolveColor(properties.listRowBackground, logger);
    }
    if (typeof properties.listRowSeparatorTint === "string") {
        style.separatorColor = resolveColor(properties.listRowSeparatorTint, logger);
    }
    style.separatorHidden = properties.listRowSeparator === "hidden";
    const insets = properties.listRowInsets;
    if (typeof insets === "number") {
        style.padding = `${insets}px`;
    } else if (typeof insets === "object" && insets !== null) {
        const { top = 0, leading = 0, bottom = 0, trailing = 0 } = insets;
        style.padding = `${top}px ${trailing}px ${bottom}px ${leading}px`;
    }
    return style;
}

function applyRowStyle(row, style) {
    if (style.background) row.style.backgroundColor = style.background;
    if (style.padding) row.style.padding = style.padding;
    if (style.separatorHidden) row.classList.add("aui-list-row-no-separator");
    if (style.separatorColor) row.style.borderBottomColor = style.separatorColor;
}
